pkg/cloud/rgraph/algo/parallel_queue.go

```diff
@@ -0,0 +1,241 @@
+		// the <-q.in will happen AFTER append(q.pending).
+		q.in <- struct{}{}
+	}
+}
+
+// Run the queue using op() to process each task. Different op()s must
+// not have interdependencies (e.g. op1 is blocked on op2 completing),
+// as this can result in a deadlock.
+//
+// Queue execution will stop if op() returns an error or the context
+// is canceled. When an error occurs, there may be goroutines that are
+// continuing to execute. Use q.WaitForOrphans() to wait for them
+// remaining goroutines.
+func (q *ParallelQueue[T]) Run(ctx context.Context, op func(context.Context, T) error) error {
+	q.lock.Lock()
+	if q.state != stateNotStarted {
+		return fmt.Errorf("Run() can only be called once (state=%d)", q.state)
+	}
+	q.state = stateRunning
+	q.lock.Unlock()
+
+	for {
+		q.lock.Lock()
+
+		klog.V(4).Infof("Run loop: pending: %d active: %d", len(q.pending), q.active)
+		if len(q.pending) == 0 && q.active == 0 {
+			q.state = stateDone
+			q.lock.Unlock()
+			return nil
+		}
+		q.launch(ctx, op)
+
+		q.lock.Unlock()
+
+		select {
+		case <-ctx.Done():
+			q.lock.Lock()
+			q.state = stateDone
+			klog.V(2).Infof("Context is Done, exiting early (pending: %d active: %d): %v", len(q.pending), q.active, ctx.Err())
+			q.lock.Unlock()
+			return ctx.Err()
+		case ri := <-q.done:
+			q.c.tracer(ri)
+			q.lock.Lock()
+			q.active--
+
+			if ri.Err != nil {
+				q.state = stateDone
+				klog.V(2).Infof("Task error, exiting early (pending: %d active: %d): %v", len(q.pending), q.active, ri.Err)
+				q.lock.Unlock()
+				return ri.Err
+			}
+
+			q.lock.Unlock()
+		case <-q.in:
+			klog.V(4).Info("<-q.in")
+			// Fall through to launch().
+		}
+	}
+}
+
+// launch task(s) if possible.
+//
+// Precondition: q.lock must be locked.
+func (q *ParallelQueue[T]) launch(ctx context.Context, op func(context.Context, T) error) {
+	pop := func() queueElement[T] {
+		t := q.pending[0]
+		q.pending = q.pending[1:]
+		return t
+	}
+
+	klog.V(4).Infof("Launch: active: %d/%d pending: %d", q.active, q.c.workerCount, len(q.pending))
+
+	for q.active < q.c.workerCount && len(q.pending) > 0 {
+		elt := pop()
+		ri := elt.ri
+		q.active++
+		klog.V(4).Infof("Launch: %q active: %d/%d pending: %d", elt.item, q.active, q.c.workerCount, len(q.pending))
+		go func() {
+			klog.V(4).Infof("Task %q start", elt.item)
+			ri.Start = time.Now()
+			ri.Err = op(ctx, elt.item)
+			ri.End = time.Now()
+			klog.V(4).Infof("Task %q end", elt.item)
+			q.done <- ri
+		}()
+	}
+}
+
+// WaitForOrphans will block until remaining op() goroutines
+// finish. Call this if Run() returns an error and you need to know
+// that all remaining threads of execution are done.
+//
+// WaitForOrphans will exit early if ctx is cancelled. This `ctx`
+// should be different from the `ctx` given to Run().
+func (q *ParallelQueue[T]) WaitForOrphans(ctx context.Context) error {
+	q.lock.Lock()
+	if q.state != stateDone {
+		return fmt.Errorf("WaitForOrphans called when not done (state = %d)", q.state)
```

Comment:
unlock

Reply:
fixing -- the change merged, so this will be a follow up PR.